Fix reversed switches false click

jeff-dagenais opened the pull request:
This is to avoid momentarily driving the pin in the incorrect level.

Brief phantom "presses" of a "reverse" GpioSwitch have been reported a
few times and this is the cause.

A good example is a GPO controlling a relay which controls a physical
button like a garage door opener remote wall control. On Pi boot, the
GpioOutput would set the mode to output, but pigpiod has the value 0
configured by default. This would result in a permanent press of the
button, which could i this example open a garage door.

Raspberry Pi's first few GPIO pins as pulled high on power on. This
means if you want to wire the Pi GPO pin directly to the relay, you must
absolutely must set the relay to active low, and use the N.O. relay
contacts to operate the button.

In this configuration (reverse and pulse), the desired "unpressed
switch" state must be programmed into PigpioD prior to setting the pin's
mode to output.

The "unpressed" depends on the "reverse" property.

jeff-dagenais committed:
GpioSwitch: set pin state before direction

This is to avoid momentarily driving the pin in the incorrect level.

Brief phantom "presses" of a "reverse" GpioSwitch have been reported a
few times and this is the cause.

A good example is a GPO controlling a relay which controls a physical
button like a garage door opener remote wall control. On Pi boot, the
GpioOutput would set the mode to output, but pigpiod has the value 0
configured by default. This would result in a permanent press of the
button, which could i this example open a garage door.

Raspberry Pi's first few GPIO pins as pulled high on power on. This
means if you want to wire the Pi GPO pin directly to the relay, you must
absolutely must set the relay to active low, and use the N.O. relay
contacts to operate the button.

In this configuration (reverse and pulse), the desired "unpressed
switch" state must be programmed into PigpioD prior to setting the pin's
mode to output.

The "unpressed" depends on the "reverse" property.

jeff-dagenais committed:
GpioSwitch: fix update function debug print

The print says "low" and "high" irrespective of the "reverse" parameter.

This is opposite to reality when "reverse" is set.

ebaauw commented:
> This has been reported a few times

Where?

> Raspberry Pi's default GPIO pins as pulled high on power on

If memory serves, that depends on the pin.  Note that pull up or pull down resisters only apply to pins in input mode.

> When GpioOutput.init occurs though, it sets the mode output of the pin prior to setting the value on the pin. 

Of course, you need to set the pin to output, before you can write a value.  By design, Homebridge RPi does **not** set the pin value on (Homebridge) startup, see #54 and #32.  You don't want the door to close (or open) unexpectedly when Homebridge restarts.

> Simple fix is to first set the right "off" value

Yes, but that should be done when the Pi boots, not when Homebridge RPi starts.
You can specify the boot values this in `config.txt`, see https://www.raspberrypi.com/documentation/computers/config_txt.html#gpio-control.

jeff-dagenais commented on lib/RpiService/GpioOutput/GpioSwitch.js:
```diff
@@ -59,8 +59,22 @@ class GpioSwitch extends GpioOutput {
+        this.warn(error)
+      }
+    }
+    return super.init()
```
In the case the `PigpioClient.commands.WRITE` fails, in the example of a pusled reverse switch controlling a door opener, I'm thinking I really don't want the `super.init()` to run! 🤔

jeff-dagenais commented:
> Did you actually ever see a WRITE fail? If so, what was the cause?

I have not. I was merely copy pasting [your suggested implementation](https://github.com/ebaauw/homebridge-rpi/pull/189#issuecomment-3279335212) actually, and the explicit handling of of the WRITE failure as a warning later made me rethink. I'll push a variation.

jeff-dagenais commented:
Alright. Updated and tested good. @ebaauw 

I removed the try/catch in `GpioSwitch.init()`. The idea is if ever this WRITE fails, we'd want to let the exception abort the rest of the `init()` implementation (i.e. calling `super.init()`) This is actually consistent with `GpioOutput.init()` where any exception is allowed to abort the function execution.

I hope you can merge this as is! I'm glad you took the time to scrutinize and push back a little. I do find this way of doing it much cleaner.

Cheers!
